### PR TITLE
APF e2e tests: add request drown-out fairness test 

### LINKED ISSUE #93757: Discussion of Priority and Fairness End-to-End Tests for Beta
## Basic CRUD

Should we create tests around the creation, modification and deletion of the `flowcontrol.apiserver.k8s.io` objects ? For example, adding an invalid FlowSchema with a misconfigured PriorityLevelConfiguration and verifying that the Status appropriately reflects its invalidity. I’m assuming the controller tests will suffice and we don’t need explicit e2e tests for this.

What about CRUD tests wherein we’re validating the debug endpoints are showing the right values ?

## Request Categorization Test

The first class of tests are the basic ones to verify that the requests are being categorized into the right FlowSchema and Priority level. We’ll set up a new PriorityLevelConfiguration and FlowSchema objects and observe that requests are correctly being classified onto them. Verification that requests are correctly being classified will be tricky. For now, we can scrape the metrics endpoint for ‘`apiserver_flowcontrol_dispatched_requests_total`’ metric to verify that the expected number of requests for a test FlowSchema is correct. 

Another option is to utilize the ‘dump\_requests’ debug endpoint to verify the correct request classification. If during the course of writing the tests we need additional information, we can extend the debug endpoint.

Scraping the metrics endpoints to validate the proper request categorization could lead to a fickle e2e test. So querying the ‘debug’ endpoint seems like the right way to go. This test will also serve as a proxy for an end-to-end test that ensures that requests matching a new FlowSchema are successful. 

#### Open Questions

*   Does this have the propensity to be a flaky test(s) ?

## Request Drown Out

This class of tests will focus on validating the behavior this feature promises, viz., requests of low priority won’t be drowned out by requests of higher priority and a single bad actor won’t detrimentally affect the requests from others. The following tests will ensure that high velocity flows don’t affect the low velocity flows.

### Priority

For this test, we’ll have two entities submitting requests at the same time. These requests will map to two different FlowSchemas at different priority levels. The higher priority flow will have a lot of requests going through it, far above its concurrency share. At the same time the low priority flow will be submitting requests as well at a rate that well within the concurrency limits of that Flow. The test will ensure that the low velocity (low priority) flow succeeds. 

### Fairness

This test will be the same as the above one except that the two flows will map to the same FlowSchema but are separated via their Flow Distinguisher. One of them will be a high velocity flow which the other is a low velocity one. The test will ensure that the low velocity requests still get through and succeed.

#### Open Questions

*   Does this have the propensity to be a flaky tests ?
*   If the test fails, will the runner of this test be able to grok where the problem could lie and debug ? Is this an actionable failure ?

### PR DESCRIPTION
**Pull request by adtac**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**: This PR adds a test to ensure that APF maintains its fairness guarantees in a request drown-out test as described in https://github.com/kubernetes/kubernetes/issues/93757.

**Which issue(s) this PR fixes**:

Fixes #93757

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

cc @lavalamp @MikeSpreitzer @yue9944882 @ahg-g 

@lavalamp - as promised in https://github.com/kubernetes/kubernetes/pull/96124#discussion_r520097400 :)

### CODE REVIEW

File: test/e2e/apimachinery/flowcontrol.go [hunk 1]
```diff
@@ -141,113 +88,101 @@ var _ = SIGDescribe("API priority and fairness", func() {
...
+
+		type client struct {
+			username              string
+			qps                   float64
+			concurrencyMultiplier float64
+			concurrency           int32
+			completedRequests     int32
+		}
+		clients := []client{
+			{username: "elephant", qps: 100.0, concurrencyMultiplier: 2.0},
```
Comment (deads2k) [inline test/e2e/apimachinery/flowcontrol.go:173]: nit: fwiw, I prefer the highqps and lowqps from above

File: test/e2e/apimachinery/flowcontrol.go [hunk 2]
```diff
@@ -271,13 +206,123 @@ var _ = SIGDescribe("API priority and fairness", func() {
 				framework.Failf("client %q: got %.1f%% completed requests, want at least 95%%", client.username, 100*fractionCompleted)
 			}
 		}
+		wg.Wait()
```
Comment (deads2k) [inline test/e2e/apimachinery/flowcontrol.go:209]: why isn't the wait on new line 197 enough?

**Comment by deads2k** [comment]
/milestone v1.20
/lgtm
/approve
/hold

hold for answer on question.  if it's needed, release the hold.  If it isn't, please update and get me on slack.

File: test/e2e/apimachinery/flowcontrol.go (hunk 2 above)
Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go:209]: ah not sure how that crept in, fixing

File: test/e2e/apimachinery/flowcontrol.go (hunk 1 above)
Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go:173]: fair, changing (elephant/mouse comes from the linked issue, but highqps/lowqps is more descriptive)

**Review by yue9944882 — COMMENTED** [review]
/lgtm

File: test/e2e/apimachinery/flowcontrol.go [hunk 3]
```diff
@@ -141,112 +88,100 @@ var _ = SIGDescribe("API priority and fairness", func() {
...
 		}
 
 		ginkgo.By("getting request concurrency from metrics")
 		for i := range clients {
-			resp, err := f.ClientSet.CoreV1().RESTClient().Get().RequestURI("/metrics").DoRaw(context.TODO())
-			framework.ExpectNoError(err)
-			sampleDecoder := expfmt.SampleDecoder{
-				Dec:  expfmt.NewDecoder(bytes.NewBuffer(resp), expfmt.FmtText),
-				Opts: &expfmt.DecodeOptions{},
+			realConcurrency := getPriorityLevelConcurrency(f, clients[i].priorityLevelName)
```
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:115]: I see that `getPriorityLevelConcurrency` just tries one scrape and fails if that scrape does not have the sought metric.  But I see no waiting to be sure the apiserver has digested the new PriorityLevelConfiguration object and started producing the concurrency metric for it.

File: test/e2e/apimachinery/flowcontrol.go [hunk 4]
```diff
@@ -141,112 +88,100 @@ var _ = SIGDescribe("API priority and fairness", func() {
...
+			clients[i].concurrency = int32(float64(realConcurrency) * clients[i].concurrencyMultiplier)
+			if clients[i].concurrency < 1 {
+				clients[i].concurrency = 1
+			}
+			framework.Logf("request concurrency for %q will be %d (concurrency share = %d)", clients[i].username, clients[i].concurrency, realConcurrency)
```
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:120]: The last quantity logged here is not a share, it is an absolute concurrency value.  "shares" are the relative values used to apportion the total allowed concurrency into absolute concurrency limits.

File: test/e2e/apimachinery/flowcontrol.go (hunk 3 above)
Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go:115]: created https://github.com/kubernetes/kubernetes/issues/96893 so that I don't forget

**Pull request merged**